theano/gpuarray/opt.py

```diff
@@ -210,6 +210,84 @@ def apply(self, fgraph):
                     0, 'fast_run', 'fast_compile', 'merge')
 
 
+class GraphToGPU(Optimizer):
+    """
+    Transfer the graph as a whole to GPU instead of transfering node by node.
+    """
+    def add_requirements(self, fgraph):
+        fgraph.attach_feature(toolbox.ReplaceValidate())
+
+    def apply(self, fgraph):
+        mapping = {}
+        move_to_GPU = True
+
+        # Building a new graph
+        # Iterating through inputs of graph
+        for i in fgraph.inputs:
+            if isinstance(i.type, tensor.TensorType):
+                mapping[i] = GpuFromHost(None)(i)
+            else:
+                mapping[i] = i
+
+        # Iterating through output of all the nodes
+        for n in fgraph.toposort():
+            for o in n.outputs:
+                if isinstance(o.type, tensor.TensorType):
+                    mapping[o] = GpuFromHost(None)(o)
+                else:
+                    mapping[o] = o
+
+        for node in fgraph.toposort():
+
+            # The Extra condition
+            if any([isinstance(i, GpuArrayVariable) or
+               isinstance(i, GpuArraySharedVariable)
+               for i in node.inputs + node.outputs]):
+
+                move_to_GPU = False
+
+            # Oplifter's condition
+            # Will return a list of OP
+            # If None, means can't be moved.
+
+            new_ops = None
+
+            # Selecting the best optimizer
+            # TODO : the tag should be updated to the one user provides
```

Comment:
To solve this, I think we should make a new DB subinstance. I would call it GraphToGPUDB(). It will receive the tags used and query the gpu_optimizer about them. and pass those 2 lists of optimization to the GraphToGPU() instance.